File: src/main/java/thedarkcolour/futuremc/recipe/StonecutterRecipes.java
package thedarkcolour.futuremc.recipe;

import com.google.common.collect.Lists;
import net.minecraft.block.Block;
import net.minecraft.init.Blocks;
import net.minecraft.item.ItemStack;
import thedarkcolour.futuremc.init.Init;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Optional;

public final class StonecutterRecipes {
    private static final ArrayList<StonecutterRecipe> RECIPES = Lists.newArrayList();

    public static void addDefaults() {
        addRecipe(Blocks.STONE, Blocks.STONE_SLAB, Blocks.STONEBRICK).addOutput(new ItemStack(Blocks.STONE_SLAB, 1, 5)).addOutput(new ItemStack(Blocks.STONE_BRICK_STAIRS)).addOutput(Init.STONE_BRICK_WALL).addOutput(new ItemStack(Blocks.STONEBRICK, 1, 3));
        addRecipe(Init.SMOOTH_STONE, new ItemStack(Blocks.STONE_SLAB, 1, 0));
        addRecipe(Blocks.STONEBRICK, Blocks.STONE_BRICK_STAIRS, Init.STONE_BRICK_WALL).addOutput(new ItemStack(Blocks.STONEBRICK, 1, 3));
        addRecipe(new ItemStack(Blocks.STONEBRICK, 1, 1), new ItemStack(Init.MOSSY_STONE_WALL));
        addRecipe(new ItemStack(Blocks.STONE, 1, 1)).addOutput(Init.GRANITE_WALL).addOutput(new ItemStack(Blocks.STONE, 1, 2));
        addRecipe(new ItemStack(Blocks.STONE, 1, 3)).addOutput(Init.DIORITE_WALL);
        addRecipe(new ItemStack(Blocks.STONE, 1, 5)).addOutput(Init.ANDESITE_WALL);
        addRecipe(Blocks.COBBLESTONE, new ItemStack(Blocks.STONE_SLAB, 1, 3), new ItemStack(Blocks.STONE_STAIRS), new ItemStack(Blocks.COBBLESTONE_WALL));
        addRecipe(new ItemStack(Blocks.MOSSY_COBBLESTONE)).addOutput(Blocks.COBBLESTONE_WALL);
        addRecipe(new ItemStack(Blocks.SANDSTONE), new ItemStack(Blocks.STONE_SLAB, 1, 1), new ItemStack(Blocks.SANDSTONE_STAIRS), new ItemStack(Init.SANDSTONE_WALL), new ItemStack(Blocks.SANDSTONE, 1, 2), new ItemStack(Blocks.SANDSTONE, 1, 1));
        addRecipe(new ItemStack(Blocks.RED_SANDSTONE), new ItemStack(Blocks.STONE_SLAB2), new ItemStack(Blocks.RED_SANDSTONE_STAIRS), new ItemStack(Init.RED_SANDSTONE_WALL), new ItemStack(Blocks.RED_SANDSTONE, 1, 2), new ItemStack(Blocks.RED_SANDSTONE, 1, 1));
        addRecipe(new ItemStack(Blocks.PRISMARINE)).addOutput(Init.PRISMARINE_WALL);
        addRecipe(new ItemStack(Blocks.QUARTZ_BLOCK), new ItemStack(Blocks.STONE_SLAB, 1, 7), new ItemStack(Blocks.QUARTZ_STAIRS), new ItemStack(Blocks.QUARTZ_BLOCK, 1, 1), new ItemStack(Blocks.QUARTZ_BLOCK, 1, 2));
        addRecipe(Blocks.PURPUR_BLOCK, Blocks.PURPUR_SLAB, Blocks.PURPUR_STAIRS, Blocks.PURPUR_PILLAR);
        addRecipe(Blocks.BRICK_BLOCK, new ItemStack(Blocks.STONE_SLAB, 1, 4)).addOutput(Blocks.BRICK_STAIRS);
        addRecipe(Blocks.NETHER_BRICK, new ItemStack(Blocks.STONE_SLAB, 1, 6), new ItemStack(Blocks.NETHER_BRICK_STAIRS), new ItemStack(Init.NETHER_BRICK_WALL));
        addRecipe(Blocks.RED_NETHER_BRICK, Init.RED_NETHER_BRICK_WALL);
        addRecipe(Blocks.END_STONE, Blocks.END_BRICKS, Init.END_STONE_WALL);
        addRecipe(Blocks.END_BRICKS, Init.END_STONE_WALL);
    }

    public static StonecutterRecipe addRecipe(ItemStack input, ItemStack... outputs) {
        StonecutterRecipe recipe = new StonecutterRecipe(input, outputs);
        RECIPES.add(recipe);
        return recipe;
    }

    public static StonecutterRecipe addRecipe(Block input, ItemStack... outputs) {
        return addRecipe(new ItemStack(input), outputs);
    }

    public static StonecutterRecipe addRecipe(Block input, Block... outputs) {
        return addRecipe(new ItemStack(input), Arrays.stream(outputs).map(ItemStack::new).toArray(ItemStack[]::new));
    }

    public static void addOrCreateRecipe(ItemStack input, ItemStack... outputs) {
        Optional<StonecutterRecipe> recipe = getRecipe(input);
        if (recipe.isPresent()) {
            for (ItemStack output : outputs) {
                recipe.get().addOutput(output);
            }
        } else {
            addRecipe(input, outputs);
        }
    }

    public static void removeOutputs(ItemStack input, ItemStack... outputs) {
        getRecipe(input).ifPresent(recipe -> {
            for (ItemStack output : outputs) {
                recipe.removeOutput(output);
            }
        });
    }

    public static Optional<StonecutterRecipe> getRecipe(ItemStack input) {
        return RECIPES.stream().filter(recipe -> recipe.matches(input)).findFirst();
    }

    public static void removeRecipe(ItemStack input) {
        RECIPES.removeIf(recipe -> recipe.matches(input));
    }

    public static void clear() {
        RECIPES.clear();
    }

    public static ArrayList<StonecutterRecipe> allRecipes() {
        return RECIPES;
    }
}

File: src/main/java/thedarkcolour/futuremc/recipe/stonecutter/StonecutterRecipes.kt
package thedarkcolour.futuremc.recipe.stonecutter

import net.minecraft.init.Blocks
import net.minecraft.item.ItemStack
import thedarkcolour.futuremc.recipe.Recipes
import thedarkcolour.futuremc.recipe.SimpleRecipe
import thedarkcolour.futuremc.registry.FBlocks

/**
 * @author TheDarkColour
 */
object StonecutterRecipes : Recipes<SimpleRecipe>() {
    override val recipes: ArrayList<SimpleRecipe>

    init {
        val stone = ItemStack(Blocks.STONE)
        val smoothStone = ItemStack(FBlocks.SMOOTH_STONE)
        val stoneBricks = ItemStack(Blocks.STONEBRICK)
        val sandstone = ItemStack(Blocks.SANDSTONE)
        val redSandstone = ItemStack(Blocks.RED_SANDSTONE)
        val andesite = ItemStack(Blocks.STONE, 1, 5)
        val diorite = ItemStack(Blocks.STONE, 1, 3)
        val granite = ItemStack(Blocks.STONE, 1, 1)
        val quartz = ItemStack(Blocks.QUARTZ_BLOCK)
        val cobblestone = ItemStack(Blocks.COBBLESTONE)
        val purpur = ItemStack(Blocks.PURPUR_BLOCK)
        val bricks = ItemStack(Blocks.BRICK_BLOCK)
        val netherBrick = ItemStack(Blocks.NETHER_BRICK)

        recipes = arrayListOf(
            SimpleRecipe(stone, ItemStack(Blocks.STONE_SLAB, 2)),
            SimpleRecipe(stone, ItemStack(Blocks.STONEBRICK)),
            SimpleRecipe(stone, ItemStack(Blocks.STONE_SLAB, 2, 5)),
            SimpleRecipe(stone, ItemStack(Blocks.STONE_BRICK_STAIRS)),
            SimpleRecipe(stone, ItemStack(FBlocks.STONE_BRICK_WALL)),
            SimpleRecipe(stone, ItemStack(Blocks.STONEBRICK, 1, 3)),
            SimpleRecipe(smoothStone, ItemStack(Blocks.STONE_SLAB, 2, 0)),
            SimpleRecipe(stoneBricks, ItemStack(Blocks.STONE_BRICK_STAIRS)),
            SimpleRecipe(stoneBricks, ItemStack(FBlocks.STONE_BRICK_WALL)),
            SimpleRecipe(stoneBricks, ItemStack(Blocks.STONEBRICK, 1, 3)),
            SimpleRecipe(ItemStack(Blocks.STONEBRICK, 1, 1), ItemStack(FBlocks.MOSSY_STONE_BRICK_WALL)),
            SimpleRecipe(ItemStack(Blocks.STONE, 1, 1), ItemStack(FBlocks.GRANITE_WALL)),
            SimpleRecipe(granite, ItemStack(Blocks.STONE, 1, 2)),
            SimpleRecipe(diorite, ItemStack(FBlocks.DIORITE_WALL)),
            SimpleRecipe(diorite, ItemStack(Blocks.STONE, 1, 4)),
            SimpleRecipe(andesite, ItemStack(FBlocks.ANDESITE_WALL)),
            SimpleRecipe(andesite, ItemStack(Blocks.STONE, 1, 6)),
            SimpleRecipe(cobblestone, ItemStack(Blocks.STONE_SLAB, 2, 3)),
            SimpleRecipe(cobblestone, ItemStack(Blocks.STONE_STAIRS)),
            SimpleRecipe(cobblestone, ItemStack(Blocks.COBBLESTONE_WALL)),
            SimpleRecipe(ItemStack(Blocks.MOSSY_COBBLESTONE), ItemStack(Blocks.COBBLESTONE_WALL, 1, 1)),
            SimpleRecipe(sandstone, ItemStack(Blocks.STONE_SLAB, 2, 1)),
            SimpleRecipe(sandstone, ItemStack(Blocks.SANDSTONE_STAIRS)),
            SimpleRecipe(sandstone, ItemStack(FBlocks.SANDSTONE_WALL)),
            SimpleRecipe(sandstone, ItemStack(Blocks.SANDSTONE, 1, 2)),
            SimpleRecipe(sandstone, ItemStack(Blocks.SANDSTONE, 1, 1)),
            SimpleRecipe(redSandstone, ItemStack(Blocks.STONE_SLAB2, 2)),
            SimpleRecipe(redSandstone, ItemStack(Blocks.RED_SANDSTONE_STAIRS)),
            SimpleRecipe(redSandstone, ItemStack(FBlocks.RED_SANDSTONE_WALL)),
            SimpleRecipe(redSandstone, ItemStack(Blocks.RED_SANDSTONE, 1, 2)),
            SimpleRecipe(redSandstone, ItemStack(Blocks.RED_SANDSTONE, 1, 1)),
            SimpleRecipe(ItemStack(Blocks.PRISMARINE), ItemStack(FBlocks.PRISMARINE_WALL)),
            SimpleRecipe(quartz, ItemStack(Blocks.STONE_SLAB, 1, 7)),
            SimpleRecipe(quartz, ItemStack(Blocks.QUARTZ_STAIRS)),
            SimpleRecipe(quartz, ItemStack(Blocks.QUARTZ_BLOCK, 1, 1)),
            SimpleRecipe(quartz, ItemStack(Blocks.QUARTZ_BLOCK, 1, 2)),
            SimpleRecipe(purpur, ItemStack(Blocks.PURPUR_SLAB, 2)),
            SimpleRecipe(purpur, ItemStack(Blocks.PURPUR_STAIRS)),
            SimpleRecipe(purpur, ItemStack(Blocks.PURPUR_PILLAR)),
            SimpleRecipe(bricks, ItemStack(Blocks.STONE_SLAB, 2, 4)),
            SimpleRecipe(bricks, ItemStack(Blocks.BRICK_STAIRS)),
            SimpleRecipe(bricks, ItemStack(FBlocks.BRICK_WALL)),
            SimpleRecipe(netherBrick, ItemStack(Blocks.STONE_SLAB, 2, 6)),
            SimpleRecipe(netherBrick, ItemStack(Blocks.NETHER_BRICK_STAIRS)),
            SimpleRecipe(netherBrick, ItemStack(FBlocks.NETHER_BRICK_WALL)),
            SimpleRecipe(ItemStack(Blocks.RED_NETHER_BRICK), ItemStack(FBlocks.RED_NETHER_BRICK_WALL)),
            SimpleRecipe(ItemStack(Blocks.END_BRICKS), ItemStack(FBlocks.END_STONE_BRICK_WALL)),
            SimpleRecipe(ItemStack(Blocks.END_STONE), ItemStack(Blocks.END_BRICKS)),
            SimpleRecipe(ItemStack(Blocks.END_STONE), ItemStack(FBlocks.END_STONE_BRICK_WALL))
        )
    }

    fun addRecipe(input: ItemStack, output: ItemStack) {
        recipes.add(SimpleRecipe(input, output))
    }

    fun removeRecipe(input: ItemStack, output: ItemStack) {
        recipes.removeIf { recipe -> recipe.matches(input, output) }
    }
}